src/serializer/ResidualHeapSerializer.js

```diff
@@ -931,13 +923,43 @@ export class ResidualHeapSerializer {
     let openingElement = t.jSXOpeningElement(identifier, (attributes: any), children.length === 0);
     let closingElement = t.jSXClosingElement(identifier);
 
-    let jsxElement = t.jSXElement(openingElement, closingElement, children, children.length === 0);
+    let reactElement = t.jSXElement(openingElement, closingElement, children, children.length === 0);
     // if there has been a bail-out, we create an inline BlockComment node before the JSX element
     if (val.$BailOutReason !== undefined) {
       // $BailOutReason contains an optional string of what to print out in the comment
-      jsxElement.leadingComments = [({ type: "BlockComment", value: `${val.$BailOutReason}` }: any)];
+      reactElement.leadingComments = [({ type: "BlockComment", value: `${val.$BailOutReason}` }: any)];
     }
-    return jsxElement;
+
+    let id = this.getSerializeObjectIdentifier(val);
+    // if we are hoisting this React element, put the assignment in the body
+    if (val.$CanHoist) {
+      // if the currentHoistedReactElements is not defined, we create it an emit the function call
+      // this should only occur once per additional function
+      if (this.hoistedReactElements === undefined) {
+        let funcId = t.identifier(this.functionNameGenerator.generate());
+        this.hoistedReactElements = {
+          id: funcId,
+          nodes: [],
+        };
+        let statement = t.expressionStatement(
+          t.logicalExpression(
+            "&&",
+            t.binaryExpression("===", id, t.unaryExpression("void", t.numericLiteral(0), true)),
+            t.callExpression(funcId, [])
+          )
+        );
+        this.emitter.emit(statement);
+      }
```

Comment:
@cblappert Prepack already has lazy initialization of objects that are referenced by multiple additional functions. Can we just reuse that instead of reinventing it? I'm concerned about getting the details right.

Reply:
It's possible but I'm not sure if its worth it. You might have to do a bit of refactoring there (if you want to create an initialization function for each lazily initialized Object instead of one function with a switch statement). Additionally, the `Referentializer` is a little more general than you need here (assumes that object identity and mutability are a concern) so you would probably also have to add another case in the code it generates. 

I think the idea of keeping all of that logic in a separate file from the serializer might help for maintainability though.

Also, if you marked all the scopes that a value is reachable from, `_getTargets` should be able to identify the common scope of all of the places its used and give you the proper scope to emit the hoisted function + variable.

Reply:
@cblappert I'll move the logic out in a separate PR if that's okay?